File: Python/SchemeInterpreter/4_CPS/main.py
#! vim:fileencoding=utf-8

import re, math, time, random, sys
#------------------------------
def printSExp(s):
    print s
#------------------------------
class Env(object):
    def __init__(self, preEnv, initFrame):
        self.preEnv = preEnv
        self.frame = initFrame
    def lookup(self, name):
        if name in self.frame:
            return self.frame[name]
        else:
            assert self.preEnv, 'Cant find variable: "%s"' % name
            return self.preEnv.lookup(name)
    def define(self, name, value):
        self.frame[name] = value
        return None
    def set(self, name, value):
        if name in self.frame:
            self.frame[name] = value
            return None
        else:
            assert self.preEnv, 'Cant find variable: "%s"' % name
            return self.preEnv.set(name, value)

def tokenize(s):
    s = re.sub(r';[^\n]*\n', '', s)
    s = s.replace('(', ' ( ').replace(')', ' ) ')
    s = s.replace('[', ' ( ').replace(']', ' ) ')
    return s.split()

def parse(tokens):
    if not tokens: 
        return None
    elif tokens[0] == '(':
        l = []
        tokens.pop(0)
        while tokens[0] != ')':
            l.append(parse(tokens))
        tokens.pop(0)
        return l
    elif re.match(r'\d+', tokens[0]):
        return int(tokens.pop(0))
    elif re.match(r'\d*\.\d+', tokens[0]):
        return float(tokens.pop(0))
    else:
        return tokens.pop(0)

def evalSequentially(env, expSeq, k):
    return eval(env, expSeq[0], k) if len(expSeq) == 1 else eval(env, expSeq[0], lambda _: evalSequentially(env, expSeq[1:], k))

def evalList(env, expList, k, result = []):
    return eval(env, expList[0], lambda v: evalList(env, expList[1:], k, result + [v])) if expList else (result, k)

def eval(env, exp, k):
    if isinstance(exp, str):
        return env.lookup(exp), k
    elif not isinstance(exp, list):
        return exp, k
    elif exp[0] == 'quote':
        (_, datum) = exp
        return datum, k
    elif exp[0] == 'if':
        (_, predExp, thenExp, elseExp) = exp
        return eval(env, predExp, lambda b: eval(env, thenExp, k) if b else eval(env, elseExp, k))
    elif exp[0] == 'lambda':
        (_, formals), expList = exp[:2], exp[2:]
        return (lambda actuals, k2: evalSequentially(Env(env, dict(zip(formals, actuals))), expList, k2)), k
    elif exp[0] == 'begin':
        _, expList = exp[0], exp[1:]
        return evalSequentially(env, expList, k)
    elif exp[0] == 'cond':
        (_, case), caseList = exp[:2], exp[2:]
        if caseList:
            casePred, caseExpList = case[0], case[1:]
            return eval(env, ['if', casePred, ['begin'] + caseExpList, ['cond'] + caseList], k)
        else:
            return evalSequentially(env, case, k)
    elif exp[0] == 'define':
        if isinstance(exp[1], str):
            (_, name, value) = exp
            return eval(env, value, lambda v: (env.define(name, v), k))
        else:
            name, formals, expList = exp[1][0], exp[1][1:], exp[2:]
            return eval(env, ['define', name, ['lambda', formals] + expList], k)
    elif exp[0] == 'set!':
        (_, name, value) = exp
        return eval(env, value, lambda v: (env.set(name, v), k))
    elif exp[0] == 'let':
        nameList, valueList, expList = [kv[0] for kv in exp[1]], [kv[1] for kv in exp[1]], exp[2:]
        return eval(env, [['lambda', nameList] + expList] + valueList, k)
    else:
        return evalList(env, exp, lambda args: args[0](args[1:], k))

def forceEval(env, exp):
    v, k = eval(env, exp, None)
    while k:
        v, k = k(v)
    return v

def funcsToCPS(funcs):
    return dict((name, lambda actuals, k, f=f: (f(*actuals), k)) for name, f in funcs.iteritems())

G = Env(False, 
        dict(
        funcsToCPS({
            '+':lambda a,b: a+b, '-':lambda a,b: a-b, '*':lambda a,b: a*b, '/':lambda a,b: a/b,
            'quotient':lambda a,b: a//b, 'remainder':lambda a,b: a%b,
            'sqr': lambda a: a*a, 'sqrt': lambda a: math.sqrt(a),
            'identity': lambda a: a,
            '=':lambda a,b: a==b, 'not':lambda a: not a,
            '<':lambda a,b: a<b, '>':lambda a,b: a>b, '<=':lambda a,b: a<=b, '>=':lambda a,b: a>=b,
            'eq':lambda a,b: a==b,
            'cons': lambda a,b: [a]+b, 'car': lambda l: l[0], 'cdr': lambda l: l[1:],
            'cadr': lambda l: l[1], 'caddr': lambda l: l[2], 'cadddr': lambda l: l[3],
            'drop': lambda l,n: l[n:], 'append': lambda a,b: a+b, 'length': lambda l: len(l), 
            'empty?': lambda l: len(l) == 0,
            'pretty-print': printSExp, 'display': lambda e: sys.stdout.write(repr(e)),
            'current-inexact-milliseconds': lambda :time.time() * 1000,
            'random': lambda a: random.randint(0, a),
            'eval': lambda e: forceEval(G, e),
            'exit': lambda : sys.exit(1),
            }).items() +
            {
            'true': True, 'false': False, 'else': True,
            'empty': [], 
            'call/cc': (lambda actuals,k: actuals[0]([(lambda actuals2,k2: (actuals2[0], k))], k)),
            }.items()
        ))

#------------------------------
tokens = tokenize(sys.stdin.read())
while True:
    s = parse(tokens)
    if not s: break
    v = forceEval(G, s)
    if v != None: printSExp(v)

File: Lua/SchemeInterpreter/4_CPS/main.lua
------------------------------
local function mapArray(f, a)
    for i = 1, #a do
        a[i] = f(a[i])
    end
    return a
end

local function cloneArray(a)
    local r = {}
    for i = 1, #a do
        r[i] = a[i]
    end
    return r
end

local function arrayFirst(a)
    return a[1]
end

local function arraySecond(a)
    return a[2]
end

local function identity(a)
    return a
end
------------------------------
local S_empty = {}
local pairTag = {}

local function S_pair(p) return type(p) == 'table' and p[3] == pairTag end
local function S_cons(a, b) return {a, b, pairTag} end
local function S_car(p) return p[1] end
local function S_cdr(p) return p[2] end

local function S_cadr(p) return p[2][1] end
local function S_caddr(p) return p[2][2][1] end
local function S_cadddr(p) return p[2][2][2][1] end


local function arrayFromAppendArraySlist(a, l)
    while l ~= S_empty do
        a[#a + 1], l = S_car(l), S_cdr(l)
    end
    return a
end
local function slistFromAppendArraySlist(a, l)
    for i = #a, 1, -1 do
        l = S_cons(a[i], l)
    end
    return l
end


local function S_append(l1, l2)
    return slistFromAppendArraySlist(
        arrayFromAppendArraySlist(arrayFromAppendArraySlist({}, l1), l2), 
        S_empty)
end
local function S_drop(p, n)
    for i = 1, n do
        p = S_cdr(p)
    end
    return p
end
local function S_length(p)
    n = 0
    while p ~= S_empty do
        p = S_cdr(p)
        n = n + 1
    end
    return n
end


local function SExp2Array(exp)
    if S_pair(exp) then
        return mapArray(SExp2Array, arrayFromAppendArraySlist({}, exp))
    else
        return exp
    end
end

local function array2SExp(t)
    if type(t) == 'table' then
        return mapArray(array2SExp, cloneArray(t))
    else
        return t
    end
end

local function printTable(t)
    if type(t) == 'table' then
        io.write('(')
        for i, v in ipairs(t) do
            if i > 1 then io.write(' ') end
            printTable(v)
        end
        io.write(')')
    else
        io.write(tostring(t))
    end
end

local function printSExp(e)
    printTable(SExp2Array(e))
    print()
end
------------------------------
local function parse(s)
    s = string.gsub(s, ';[^\n]*\n', '')
    s = string.gsub(s, '[%(%)%[%]]', {['(']='{',[')']='} ',['[']='{',[']']='} '})
    s = string.gsub(s, '%s+', ',')
    s = string.gsub(s, '[^{},%d][^{},]*', '"%1"')
    return assert(loadstring(string.format("return {%s}", s)))()
end

local evalSequentially
local evalList

local function eval(env, exp, k)
    local expType = type(exp)
    if expType == 'string' then
        return k(env[exp])
    elseif expType ~= 'table' then
        return k(exp)
    elseif exp[1] == 'quote' then
        return k(array2SExp(exp[2]))
    elseif exp[1] == 'if' then
        return eval(env, exp[2], function(b) 
            if b then 
                return eval(env, exp[3], k)
            else 
                return eval(env, exp[4], k) 
            end
        end)
    elseif exp[1] == 'lambda' then
        return k(function(actuals, k)
            local newEnv = {}
            for i = 1, #exp[2] do
                newEnv[exp[2][i]] = actuals[i]
            end
            setmetatable(newEnv, {__index = env, __newindex = env})
            return evalSequentially(newEnv, exp, 3, k)
        end)
    elseif exp[1] == 'begin' then
        return evalSequentially(env, exp, 2, k)
    elseif exp[1] == 'cond' then
        if #exp == 2 then
            return evalSequentially(env, exp[2], 1, k)
        else
            local caseBody = cloneArray(exp[2])
            caseBody[1] = 'begin'
            local restCase = cloneArray(exp)
            table.remove(restCase, 2)
            local newExp = {'if', exp[2][1], caseBody, restCase}
            return eval(env, newExp, k)
        end
    elseif exp[1] == 'define' then
        if type(exp[2]) == 'string' then
            return eval(env, exp[3], function(v)
                rawset(env, exp[2], v)
                return k(nil)
            end)
        else
            local newExp = cloneArray(exp)
            newExp[1] = 'lambda'
            newExp[2] = cloneArray(newExp[2])
            table.remove(newExp[2], 1)
            newExp = {'define', exp[2][1], newExp}
            return eval(env, newExp, k)
        end
    elseif exp[1] == 'set!' then
        return eval(env, exp[3], function(v)
            env[exp[2]] = v
            return k(nil)
        end)
    elseif exp[1] == 'let' then
        local newExp = {'lambda', mapArray(arrayFirst, cloneArray(exp[2])) }
        for i = 3, #exp do
            newExp[#newExp + 1] = exp[i]
        end
        newExp = {newExp}
        for _, v in ipairs(mapArray(arraySecond, cloneArray(exp[2]))) do
            newExp[#newExp + 1] = v
        end
        return eval(env, newExp, k)
    else
        return eval(env, exp[1], function(f)
            return evalList(env, exp, 2, function(actuals)
                return f(actuals, k)
            end)
        end)
    end
end

evalSequentially = function(env, seq, start, k)
    if start == #seq then
        return eval(env, seq[start], k)
    else
        return eval(env, seq[start], function(_)
            return evalSequentially(env, seq, start + 1, k)
        end)
    end
end

evalList = function(env, list, start, k, result, i)
    result = result or {}
    i = i or 1
    if start > #list then
        return k(result)
    else
        return eval(env, list[start], function(v)
            result[i] = v
            return evalList(env, list, start + 1, k, result, i + 1)
        end)
    end
end

local function wrapFuncToCPS(f)
    return function(actuals, k)
        return k(f(unpack(actuals)))
    end
end

local G = {
    ['+'] = function(a, b) return a + b end, ['-'] = function(a, b) return a - b end,
    ['*'] = function(a, b) return a * b end, ['/'] = function(a, b) return a / b end,
    ['quotient'] = function(a, b) return math.floor(a / b) end, ['remainder'] = function(a, b) return a % b end,
    ['sqr'] = function(a) return a * a end, ['sqrt'] = function(a) return math.sqrt(a) end,
    ['identity'] = identity,
    ['true'] = true, ['false'] = false, ['else'] = true,
    ['='] = function(a, b) return a == b end, ['not'] = function(a) return not a end,
    ['<'] = function(a, b) return a < b end, ['<='] = function(a, b) return a <= b end,
    ['>'] = function(a, b) return a > b end, ['>='] = function(a, b) return a >= b end,
    ['cons'] = S_cons, ['car'] = S_car, ['cdr'] = S_cdr,
    ['cadr'] = S_cadr, ['caddr'] = S_caddr, ['cadddr'] = S_cadddr, 
    ['drop'] = S_drop, ['append'] = S_append, ['length'] = S_length, 
    ['empty'] = S_empty, ['empty?'] = function(a) return a == S_empty end,
    ['pretty-print'] = printSExp, ['display'] = function(e) return io.write(tostring(e)) end,
    ['current-inexact-milliseconds'] = function() return os.clock() * 1000 end,
    ['random'] = function(a) return math.random(a) - 1 end,
    ['exit'] = os.exit,
}
G['eval'] = function(e) return eval(G, SExp2Array(e), identity) end

for k, v in pairs(G) do
    if type(v) == 'function' then
        v = wrapFuncToCPS(v)
    end
    G[k] = v
end

G['call/cc'] = function(actuals, k) 
    return actuals[1]({
        function(actuals2, k2) 
            return k(actuals2[1])
        end}, 
    k) 
end
------------------------------
for _, e in ipairs(parse(io.read('*a'))) do
    local v = eval(G, e, identity)
    if v ~= nil then printSExp(v) end
end
